File: feature/shiftgen/core/parser.py
import re
from datetime import datetime
from typing import List, Dict, Tuple

from bs4 import BeautifulSoup


class ScheduleParser:
    """Parser for ShiftGen schedule HTML"""
    
    @staticmethod
    def determine_role_from_site(site_name: str) -> str:
        """
        Determine the role based on the site name.
        
        Args:
            site_name: Name of the site
            
        Returns:
            Role string (Scribe, Physician, or MLP)
        """
        site_lower = site_name.lower()
        if "scribe" in site_lower:
            return "Scribe"
        elif "physician" in site_lower:
            return "Physician"
        elif "mlp" in site_lower:
            return "MLP"
        else:
            return "Unknown"
    
    @staticmethod
    def parse_shift_text(shift_text: str, role: str = "") -> Tuple[str, str, str]:
        """
        Parse shift text into components.
        
        Args:
            shift_text: Raw shift text from HTML
            role: Role type for context
            
        Returns:
            Tuple of (label, time, person)
        """
        if not shift_text:
            return "", "", ""

        # Normalize whitespace
        s = shift_text.strip()
        s = s.replace("\u00A0", " ").replace("\u200b", "")
        s = s.replace("\r", " ").replace("\n", " ").replace("\t", " ")
        s = re.sub(r"\s+", " ", s).strip()
        s = re.sub(r"\s*:\s*", ": ", s)

        # SJH physician prefix (e.g., "SJH A 0530-1400: MERJANIAN")
        m = re.match(r"^(?:SJH)\s+([A-Za-z0-9\- ]+?)\s+(\d{3,4}-\d{3,4}):\s*(.+)$", s, flags=re.IGNORECASE)
        if m:
            return m.group(1).strip(), m.group(2), m.group(3).strip()

        # CHOC physician directions (e.g., "North 0530-1400: SHIEH")
        m = re.match(r"^(North|South|East|West|RED)\s+(\d{3,4}-\d{3,4}):\s*(.+)$", s, flags=re.IGNORECASE)
        if m:
            return m.group(1).strip(), m.group(2), m.group(3).strip()

        # CHOC MLP entries -> normalize to "PA"
        m = re.match(r"^CHOC\s+(?:MLP|PA|[A-Za-z0-9\- ]+?)\s+(\d{3,4}-\d{3,4}):\s*(.+)$", s, flags=re.IGNORECASE)
        if m:
            return "PA", m.group(1), m.group(2).strip()

        # General "Label TIME: Person"
        m = re.match(r"^([A-Za-z0-9\- ]{1,30}?)\s+(\d{3,4}-\d{3,4}):\s*(.+)$", s)
        if m:
            return m.group(1).strip(), m.group(2), m.group(3).strip()

        # Time + Role pattern (e.g., "1000-1830 PA: Molly")
        m = re.match(r"^(\d{3,4}-\d{3,4})\s*(PA|MD|NP|RN):\s*(.+)$", s, flags=re.IGNORECASE)
        if m:
            return m.group(2).upper(), m.group(1), m.group(3).strip()

        # Time (Location) : Person (e.g., "1000-1800 (RED): Ahilin")
        m = re.match(r"^(\d{3,4}-\d{3,4})\s*\(([^)]+)\):\s*(.+)$", s)
        if m:
            return m.group(2).strip(), m.group(1), m.group(3).strip()

        # Simple "TIME: Person"
        m = re.match(r"^(\d{3,4}-\d{3,4}):\s*(.+)$", s)
        if m:
            return "", m.group(1), m.group(2).strip()

        # Fallback
        if ":" in s:
            left, right = s.split(":", 1)
            left = left.strip()
            person = right.strip()
            time_match = re.search(r"(\d{3,4}-\d{3,4})", left)
            if time_match:
                time = time_match.group(1)
                label_part = left[:left.index(time)].strip()
                label = re.sub(r"\b(SJH|CHOC)\b", "", label_part, flags=re.IGNORECASE).strip()
                return label, time, person

        return "", "", s
    
    @staticmethod
    def normalize_person(person: str) -> str:
        """
        Normalize person name.
        
        Args:
            person: Raw person string
            
        Returns:
            Normalized person name
        """
        person = person.strip()
        if "**EMPTY**" in person.upper() or "EMPTY" in person.upper():
            return "EMPTY"
        return person
    
    def parse_calendar(self, html_content: str, site_name: str = "") -> List[Dict[str, str]]:
        """
        Parse calendar HTML into structured data.
        
        Args:
            html_content: HTML content of the schedule
            site_name: Name of the site for role determination
            
        Returns:
            List of shift dictionaries
        """
        soup = BeautifulSoup(html_content, "html.parser")
        
        # Extract month/year from header
        header = soup.find("div", style=lambda x: x and "font-weight:bold" in x and "font-size:16px" in x)
        if not header:
            return []
        
        header_text = header.get_text(strip=True)
        month_year = None
        
        # Try to extract month and year using regex
        match = re.search(r"([A-Za-z]+\s+\d{4})", header_text)
        if match:
            month_year = match.group(1)
        
        if not month_year:
            # Fallback: try to extract from the date range in parentheses
            match = re.search(r"\((\d{2}/\d{2}/\d{4})", header_text)
            if match:
                date_str = match.group(1)
                try:
                    date_obj = datetime.strptime(date_str, "%m/%d/%Y")
                    month_year = date_obj.strftime("%B %Y")
                except ValueError:
                    return []
            else:
                return []
        
        role = self.determine_role_from_site(site_name)
        schedule_data = []
        
        # Find all day cells
        for day_cell in soup.find_all("td", style=lambda x: x and "vertical-align:text-top" in x):
            day_div = day_cell.find("div", style=lambda x: x and "font-size:12px" in x)
            if not day_div:
                continue
            
            day_num = day_div.get_text(strip=True)
            if not day_num.isdigit():
                continue
            
            # Construct date
            try:
                date_obj = datetime.strptime(f"{day_num} {month_year}", "%d %B %Y")
                date_str = date_obj.strftime("%Y-%m-%d")
            except ValueError:
                continue
            
            # Extract shifts (ignore notes in <pre>)
            for span in day_cell.find_all("span"):
                shift_text = span.get_text(strip=True)
                if not shift_text:
                    continue
                
                label, time, person = self.parse_shift_text(shift_text, role)
                person = self.normalize_person(person)
                
                # Skip empty shifts
                if person != "EMPTY":
                    schedule_data.append({
                        "date": date_str,
                        "label": label.strip(),
                        "time": time.strip(),
                        "person": person,
                        "role": role,
                        "site": site_name
                    })
        
        return schedule_data

File: src/lib/shiftgen/parser.ts
/**
 * ShiftGen Schedule Parser
 *
 * Parses the ShiftGen calendar UI into structured shift records.
 *
 * ShiftGen migrated from the legacy table-based markup (inline styles,
 * `<td style="vertical-align:text-top">`, `<span>Label TIME: Person</span>`) to a
 * Rails/Hotwire + Tailwind app. The old markup no longer exists anywhere on the
 * site, so this parser targets the current DOM exclusively.
 *
 * Every shift is now a single element carrying its data in attributes:
 *
 *   <div id="2026_08_01_16271"
 *        data-controller="shift-cell-component"
 *        data-shift-cell-component-shift-key-value="2026_08_01_16271"
 *        data-shift-cell-component-name-value="B 0530-1400"
 *        data-shift-cell-component-assignee-value="Isaac"
 *        data-shift-cell-component-assignee-id-value="14187">
 *
 * That means no regex against rendered text: the date, shift id, label, time and
 * assignee all come straight off the element.
 */

import * as cheerio from 'cheerio';
import { getShiftLetterFromTime } from './constants';

/** Labels used by the St Joseph scribe schedule. */
const SJH_LABELS = new Set([
  'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'X', 'PIT',
]);

/** Labels used by the CHOC schedule (directional zones). */
const CHOC_LABELS = new Set([
  'NORTH', 'SOUTH', 'EAST', 'WEST', 'RED',
]);

/** Labels that appear on both schedules and so cannot identify a site. */
const AMBIGUOUS_LABELS = new Set(['PA', 'MLP']);

export interface RawShiftData {
  date: string;      // YYYY-MM-DD
  label: string;     // Normalized zone identifier (A, B, C, PA, North, ...)
  rawLabel: string;  // Label exactly as ShiftGen wrote it (for debugging)
  time: string;      // HHMM-HHMM
  person: string;    // Assignee name, or 'EMPTY' for an unfilled shift
  role: string;      // Scribe, Physician, or MLP
  site: string;      // Site name
  shiftId?: string;      // ShiftGen's stable shift id
  scheduleId?: string;   // ShiftGen's schedule id (one per site per month)
  workerId?: string;     // ShiftGen's worker id for the assignee
  assigned: boolean;     // false when the shift is open/unfilled
  providerName?: string; // Populated during matching
  providerRole?: string; // Populated during matching
}

/** A shift whose name ShiftGen stored in a format we cannot turn into a time. */
export interface MalformedShift {
  date: string;
  name: string;
  person: string;
  reason: string;
}

interface CellData {
  date: string;
  rawName: string;
  person: string;
  assigned: boolean;
  shiftId?: string;
  scheduleId?: string;
  workerId?: string;
  siteTag?: string;
}

export class ScheduleParser {
  /**
   * Shifts dropped by the most recent parseCalendar() call because ShiftGen
   * stored a name we cannot derive a start/end time from. Surfaced rather than
   * swallowed so a sync can report them instead of silently losing coverage.
   */
  public malformedShifts: MalformedShift[] = [];

  /**
   * Determine role from site name
   */
  static determineRoleFromSite(siteName: string): string {
    const siteLower = siteName.toLowerCase();
    if (siteLower.includes('scribe')) {
      return 'Scribe';
    } else if (siteLower.includes('physician')) {
      return 'Physician';
    } else if (siteLower.includes('mlp')) {
      return 'MLP';
    } else {
      return 'Unknown';
    }
  }

  /**
   * Split a shift name into its label and time.
   *
   * ShiftGen writes the label before or after the time, and sometimes wraps it
   * in parentheses:
   *   "A 0530-1400"     -> { label: 'A',   time: '0530-1400' }
   *   "1000-1830 PA"    -> { label: 'PA',  time: '1000-1830' }
   *   "1800-0200 (RED)" -> { label: 'RED', time: '1800-0200' }
   *
   * Returns null for names with no full HHMM-HHMM time. Those are split-shift
   * entries typed by hand into ShiftGen's name field ("South 05-09",
   * "2000 PA 12-430", "D 1100 - 11-3") whose real hours are ambiguous, so we
   * refuse to guess.
   */
  static parseShiftName(shiftName: string): { label: string; time: string } | null {
    if (!shiftName) {
      return null;
    }

    // Normalize whitespace and stray unicode.
    const s = shiftName
      .replace(/ /g, ' ')
      .replace(/​/g, '')
      .replace(/\s+/g, ' ')
      .trim();

    // Require a complete HHMM-HHMM (or HMM-HMM) time with no internal spaces.
    const m = s.match(/^(.*?)\s*(\d{3,4})-(\d{3,4})\s*(.*)$/);
    if (!m) {
      return null;
    }

    const [, before, start, end, after] = m;
    const label = (before || after || '')
      .replace(/[()]/g, '')
      .trim();

    return {
      label,
      time: `${start.padStart(4, '0')}-${end.padStart(4, '0')}`,
    };
  }

  /**
   * Normalize person name
   */
  static normalizePerson(person: string): string {
    const personTrimmed = (person || '').trim();
    if (
      !personTrimmed ||
      personTrimmed.toUpperCase().includes('**EMPTY**') ||
      personTrimmed.toUpperCase() === 'EMPTY'
    ) {
      return 'EMPTY';
    }
    return personTrimmed;
  }

  /**
   * Infer which site a group of shifts belongs to from its label vocabulary.
   *
   * ShiftGen's schedule ids change every month, so they can group cells but
   * cannot name a site on their own. The label vocabulary is stable: the SJH
   * scribe schedule uses letters, CHOC uses compass directions plus RED.
   */
  static inferSiteFromLabels(labels: string[]): string | null {
    let sjh = 0;
    let choc = 0;

    for (const label of labels) {
      const upper = label.toUpperCase();
      if (AMBIGUOUS_LABELS.has(upper)) continue;
      if (SJH_LABELS.has(upper)) sjh++;
      else if (CHOC_LABELS.has(upper)) choc++;
    }

    if (sjh === 0 && choc === 0) return null;
    return sjh >= choc ? 'St Joseph Scribe' : 'CHOC Scribe';
  }

  /**
   * Extract every shift cell from the page, without interpreting it yet.
   */
  private extractCells(html: string): CellData[] {
    const $ = cheerio.load(html);
    const cells: CellData[] = [];

    $('div[data-controller~="shift-cell-component"]').each((_, el) => {
      const $cell = $(el);
      const prefix = 'data-shift-cell-component-';

      const shiftKey =
        $cell.attr(`${prefix}shift-key-value`) || $cell.attr('id') || '';

      // shift key is YYYY_MM_DD_<shiftId>
      const keyMatch = shiftKey.match(/^(\d{4})_(\d{2})_(\d{2})_(\d+)$/);
      if (!keyMatch) {
        return;
      }
      const date = `${keyMatch[1]}-${keyMatch[2]}-${keyMatch[3]}`;
      const shiftId = keyMatch[4];

      const assigneeId = $cell.attr(`${prefix}assignee-id-value`) || '';
      const person = ScheduleParser.normalizePerson(
        $cell.attr(`${prefix}assignee-value`) || ''
      );
      const assigned = person !== 'EMPTY' && assigneeId !== 'noworker';

      // The schedule id is only exposed through the shift-action links, e.g.
      // /changes/remote_request_take/2026_08_01_12895/15963/17651/12895
      let scheduleId: string | undefined;
      let workerId: string | undefined;
      $cell.find('[data-view-all-sites-url-param]').each((__, link) => {
        if (scheduleId) return;
        const url = $(link).attr('data-view-all-sites-url-param') || '';
        const m = url.match(
          /\/changes\/remote_request_\w+\/\d{4}_\d{2}_\d{2}_\d+\/(\d+)\/(\d+)\//
        );
        if (m) {
          scheduleId = m[1];
          workerId = m[2];
        }
      });

      // Single-site views render the site as a chip inside the cell.
      const siteTag = $cell
        .find('span[class*="text-tag"]')
        .first()
        .text()
        .trim();

      cells.push({
        date,
        rawName: $cell.attr(`${prefix}name-value`) || '',
        person,
        assigned,
        shiftId,
        scheduleId,
        workerId,
        siteTag: siteTag || undefined,
      });
    });

    return cells;
  }

  /**
   * Parse a ShiftGen calendar page into structured shift data.
   *
   * @param htmlContent  Raw HTML of a schedule page.
   * @param siteName     Optional site name. Supply it when scraping a
   *                     single-site page; on the multi-site view the site is
   *                     resolved per schedule instead.
   */
  parseCalendar(htmlContent: string, siteName: string = ''): RawShiftData[] {
    this.malformedShifts = [];

    const cells = this.extractCells(htmlContent);
    if (cells.length === 0) {
      return [];
    }

    // Resolve a site per schedule group, so the multi-site view attributes each
    // shift correctly and ambiguous labels (PA) inherit their schedule's site.
    const labelsBySchedule = new Map<string, string[]>();
    for (const cell of cells) {
      const parsed = ScheduleParser.parseShiftName(cell.rawName);
      if (!parsed) continue;
      const key = cell.scheduleId || '';
      if (!labelsBySchedule.has(key)) labelsBySchedule.set(key, []);
      labelsBySchedule.get(key)!.push(parsed.label);
    }

    const siteBySchedule = new Map<string, string>();
    for (const [key, labels] of labelsBySchedule) {
      const inferred = ScheduleParser.inferSiteFromLabels(labels);
      siteBySchedule.set(key, inferred || siteName || 'Unknown');
    }

    const scheduleData: RawShiftData[] = [];

    for (const cell of cells) {
      const parsed = ScheduleParser.parseShiftName(cell.rawName);

      if (!parsed) {
        this.malformedShifts.push({
          date: cell.date,
          name: cell.rawName,
          person: cell.person,
          reason: 'No parseable HHMM-HHMM time in shift name',
        });
        continue;
      }

      // Skip unfilled shifts, matching the previous parser's behaviour.
      if (!cell.assigned) {
        continue;
      }

      // A site chip on the cell is authoritative; otherwise fall back to the
      // schedule's inferred site, then to the caller-supplied name.
      const site =
        cell.siteTag ||
        siteBySchedule.get(cell.scheduleId || '') ||
        siteName ||
        'Unknown';

      // getShiftLetterFromTime() only knows the SJH letter scheme, so applying
      // it to CHOC would rewrite "North 0530-1400" as "A" and collide with the
      // real SJH A shift. Only normalize where the scheme applies.
      const isSjh = SJH_LABELS.has(parsed.label.toUpperCase());
      const isAmbiguous = AMBIGUOUS_LABELS.has(parsed.label.toUpperCase());
      const [startTime] = parsed.time.split('-');

      const label =
        isSjh || isAmbiguous
          ? getShiftLetterFromTime(startTime, parsed.label) || parsed.label
          : parsed.label;

      scheduleData.push({
        date: cell.date,
        label: label.trim(),
        rawLabel: parsed.label.trim(),
        time: parsed.time,
        person: cell.person,
        role: ScheduleParser.determineRoleFromSite(site),
        site,
        shiftId: cell.shiftId,
        scheduleId: cell.scheduleId,
        workerId: cell.workerId,
        assigned: cell.assigned,
      });
    }

    return scheduleData;
  }
}
